### PR TITLE
Fix DISTINCT ON expressions match ORDER BY expressions check - normal…

### LINKED ISSUE #13038: Distinct on expressions mismatch ORDER BY expressions because of a qualified column
### Describe the bug

When planning a query with distinct on, the distinct on expressions must match the order by expressions.
There is a check for that, however the check checks if the expressions are equal without normalizing column names first.
Because of that planning queries like "SELECT DISTINCT ON (t.a) * FROM t ORDER BY a" won't work while ""SELECT DISTINCT ON (t.a) * FROM t ORDER BY t.a" will.

### To Reproduce

Add the following to plan.rs unit tests and run:

```
#[test]
    fn distinct_on_expr_order_by_match() -> Result<()> {
        let schema = Schema::new(vec![Field::new("a", DataType::Int32, false)]);

        let table_scan =
            table_scan(Some(TableReference::bare("t")), &schema, None)?.build()?;
        let p = DistinctOn::try_new(
            vec![col("a")],
            vec![],
            Some(vec![SortExpr::new(col("t.a"), true, false)]),
            Arc::new(table_scan),
        )?;
        Ok(())
    }
```

### Expected behavior

Test should pass

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by epsio-banay**
…ize columns before checking if expressions match

## Which issue does this PR close?

Closes #13038.

## Are these changes tested?

Yes, unit tested

## Are there any user-facing changes?

No

### CODE REVIEW

File: datafusion/expr/src/logical_plan/plan.rs [hunk 1]
```diff
@@ -3870,6 +3873,39 @@ digraph {
         );
     }
 
+    #[test]
```
Comment (alamb) [inline datafusion/expr/src/logical_plan/plan.rs:3876]: Could you also please add a sqllogictest? Perhaps by extending https://github.com/apache/datafusion/blob/main/datafusion/sqllogictest/test_files/distinct_on.slt ?

File: datafusion/expr/src/logical_plan/plan.rs [hunk 2]
```diff
@@ -3870,6 +3873,39 @@ digraph {
...
+            vec![],
+            Some(vec![SortExpr::new(col("b"), true, false)]),
+            Arc::new(table_scan),
+        );
+        assert_eq!(p.err().unwrap().strip_backtrace(), "Error during planning: SELECT DISTINCT ON expressions must match initial ORDER BY expressions");
+        Ok(())
+    }
+
+    #[test]
+    fn distinct_on_expr_order_by_match() -> Result<()> {
```
Comment (alamb) [inline datafusion/expr/src/logical_plan/plan.rs:3895]: When I revert the code changes in this PR, these tests still pass 🤔 -- so I am not sure this PR solves the issue


```diff
diff --git a/datafusion/expr/src/logical_plan/plan.rs b/datafusion/expr/src/logical_plan/plan.rs
index 6f4e78ccf..9d7d893be 100644
--- a/datafusion/expr/src/logical_plan/plan.rs
+++ b/datafusion/expr/src/logical_plan/plan.rs
@@ -28,8 +28,7 @@ use super::DdlStatement;
 use crate::builder::{change_redundant_column, unnest_with_options};
 use crate::expr::{Placeholder, Sort as SortExpr, WindowFunction};
 use crate::expr_rewriter::{
-    create_col_from_scalar_expr, normalize_col, normalize_cols, normalize_sorts,
-    NamePreserver,
+    create_col_from_scalar_expr, normalize_cols, normalize_sorts, NamePreserver,
 };
 use crate::logical_plan::display::{GraphvizVisitor, IndentVisitor};
 use crate::logical_plan::extension::UserDefinedLogicalNode;
@@ -2890,9 +2889,7 @@ impl DistinctOn {
         // Check that the left-most sort expressions are the same as the `ON` expressions.
         let mut matched = true;
         for (on, sort) in self.on_expr.iter().zip(sort_expr.iter()) {
-            if normalize_col(on.clone(), &self.input)?
-                != normalize_col(sort.expr.clone(), &self.input)?
-            {
+            if on != &sort.expr {
                 matched = false;
                 break;
             }
```


```shell
(venv) andrewlamb@Andrews-MacBook-Pro-2:~/Software/datafusion$ cargo test -p datafusion-expr -- distinct_on
   Compiling datafusion-expr v42.1.0 (/Users/andrewlamb/Software/datafusion/datafusion/expr)
warning: unused variable: `p`
    --> datafusion/expr/src/logical_plan/plan.rs:3897:13
     |
3897 |         let p = DistinctOn::try_new(
     |             ^ help: if this is intentional, prefix it with an underscore: `_p`
     |
     = note: `#[warn(unused_variables)]` on by default

warning: `datafusion-expr` (lib test) generated 1 warning
    Finished `test` profile [unoptimized + debuginfo] target(s) in 2.99s
     Running unittests src/lib.rs (target/debug/deps/datafusion_expr-3be1ac638c9d6fb5)

running 2 tests
test logical_plan::plan::tests::distinct_on_expr_order_by_match ... ok
test logical_plan::plan::tests::distinct_on_expr_order_by_mismatch ... ok

test result: ok. 2 passed; 0 failed; 0 ignored; 0 measured; 107 filtered out; finished in 0.00s

   Doc-tests datafusion_expr

running 0 tests

test result: ok. 0 passed; 0 failed; 0 ignored; 0 measured; 45 filtered out; finished in 0.01s
```

**Review by alamb — COMMENTED** [review]
Thank you for this contribution @epsio-banay

File: datafusion/expr/src/logical_plan/plan.rs [hunk 3]
```diff
@@ -2889,7 +2890,9 @@ impl DistinctOn {
         // Check that the left-most sort expressions are the same as the `ON` expressions.
         let mut matched = true;
         for (on, sort) in self.on_expr.iter().zip(sort_expr.iter()) {
-            if on != &sort.expr {
+            if normalize_col(on.clone(), &self.input)?
```
[the remark this answers is not in the document]
Comment (epsio-banay) [reply datafusion/expr/src/logical_plan/plan.rs]: Sorry for wasting your time on these tests and thank you for looking into it.
Apparently the bug I experienced was just a little bit different - there is a bug in column normalization where we don't normalize field qualifiers.

for the query "SELECT DISTINCT ON (t.a) t.a AS alert_id FROM public.t ORDER BY t.a"
the sort expression was normalized to public.t.a but the distinct on expression was normalized to t.a.

I added a fix for that in the `normalize_with_schemas_and_ambiguity_check` function.
But I think this PR will now probably effect most of the logical plans that datafusion will generate.

File: datafusion/expr/src/logical_plan/plan.rs (hunk 2 above)
Comment (epsio-banay) [reply datafusion/expr/src/logical_plan/plan.rs:3892]: I changed the tests.
marking as resolved

**Pull request closed without merging**